File: bindings/r/tests/local_implementation_registry.R
library(dagml)

if (!requireNamespace("jsonlite", quietly = TRUE)) {
  stop("jsonlite is required for DAG-ML R binding tests", call. = FALSE)
}

fixture_path <- system.file(
  "extdata",
  "r_local_implementations.v1.json",
  package = "dagml"
)
stopifnot(nzchar(fixture_path))
fixture <- jsonlite::fromJSON(fixture_path, simplifyVector = FALSE)

calls <- 0L
asymmetric_loss <- function(target, prediction) {
  calls <<- calls + 1L
  difference <- prediction - target
  mean(ifelse(difference >= 0, difference^2, 2 * difference^2))
}
bias_metric <- function(target, prediction) mean(prediction - target)

registry <- dagml_local_implementation_registry()
stopifnot(inherits(registry, "dagml_local_implementation_registry"))
registry$register_loss(fixture$loss_reference, asymmetric_loss)
registry$register_metric(fixture$metric_reference, bias_metric)
stopifnot(registry$size() == 2L)
stopifnot(length(registry$descriptors()) == 2L)

for (phase in c("FIT_CV", "REFIT")) {
  invocation <- registry$invoke_training_loss(
    fixture$task_json[[phase]],
    target = c(2, 4),
    prediction = c(5, 3)
  )
  stopifnot(isTRUE(all.equal(invocation$value, 5.5)))
  stopifnot(identical(invocation$attestation$phase, phase))
  stopifnot(identical(
    invocation$attestation$descriptor_fingerprint,
    fixture$loss_reference$implementation$descriptor_fingerprint
  ))
  stopifnot(isTRUE(all.equal(
    invocation$attestation,
    fixture$tasks[[phase]]$required_loss_attestations[[1]]
  )))
}
stopifnot(calls == 2L)

stopifnot(isTRUE(all.equal(
  registry$invoke_metric(
    fixture$metric_reference,
    target = c(2, 4),
    prediction = c(5, 3)
  ),
  1
)))

error <- tryCatch(
  registry$invoke_training_loss(
    fixture$invalid_task_json$predict,
    target = 2,
    prediction = 5
  ),
  error = identity
)
stopifnot(inherits(error, "error"), calls == 2L)

error <- tryCatch(
  registry$register_loss(fixture$foreign_loss_reference, asymmetric_loss),
  error = identity
)
stopifnot(inherits(error, "error"))

portable_builtin <- fixture$loss_reference
portable_builtin$implementation$portability <- "portable_builtin"
portable_registry <- dagml_local_implementation_registry()
error <- tryCatch(
  portable_registry$register_loss(portable_builtin, asymmetric_loss),
  error = identity
)
stopifnot(inherits(error, "error"), grepl("portable_builtin", conditionMessage(error)))

error <- tryCatch(
  registry$register_loss(fixture$loss_reference, asymmetric_loss),
  error = identity
)
stopifnot(inherits(error, "error"), grepl("duplicate", conditionMessage(error)))

error <- tryCatch(
  registry$register_loss(fixture$loss_reference, 42),
  error = identity
)
stopifnot(inherits(error, "error"), grepl("R functions", conditionMessage(error)))

error <- tryCatch(
  registry$resolve_metric(fixture$loss_reference),
  error = identity
)
stopifnot(inherits(error, "error"), grepl("expected a metric", conditionMessage(error)))

drifted <- fixture$loss_reference
drifted$implementation$implementation_version <- "2.0.0"
error <- tryCatch(registry$resolve_loss(drifted), error = identity)
stopifnot(inherits(error, "error"))

error <- tryCatch(
  registry$invoke_training_loss(
    fixture$invalid_task_json$tampered_attestation,
    target = 2,
    prediction = 5
  ),
  error = identity
)
stopifnot(
  inherits(error, "error"),
  grepl("requirements that do not match", conditionMessage(error)),
  calls == 2L
)

error <- tryCatch(
  registry$invoke_training_loss(
    fixture$invalid_task_json$wrong_attestation_schema,
    target = 2,
    prediction = 5
  ),
  error = identity
)
stopifnot(
  inherits(error, "error"),
  grepl("requirements that do not match", conditionMessage(error)),
  calls == 2L
)

error <- tryCatch(
  registry$invoke_training_loss(
    fixture$invalid_task_json$missing_attestation,
    target = 2,
    prediction = 5
  ),
  error = identity
)
stopifnot(
  inherits(error, "error"),
  grepl("requirements that do not match", conditionMessage(error)),
  calls == 2L
)

error <- tryCatch(
  registry$invoke_training_loss(
    fixture$task_json$FIT_CV,
    role_index = 2L,
    target = 2,
    prediction = 5
  ),
  error = identity
)
stopifnot(
  inherits(error, "error"),
  grepl("outside the active", conditionMessage(error)),
  calls == 2L
)

without_native <- dagml_local_implementation_registry(native_library = "")
without_native$register_loss(fixture$loss_reference, asymmetric_loss)
error <- tryCatch(
  without_native$invoke_training_loss(
    fixture$task_json$FIT_CV,
    target = 2,
    prediction = 5
  ),
  error = identity
)
stopifnot(
  inherits(error, "error"),
  grepl("DAGML_NATIVE_LIBRARY", conditionMessage(error)),
  calls == 2L
)

invalid_native <- dagml_local_implementation_registry(
  native_library = "/dagml/does/not/exist"
)
invalid_native$register_loss(fixture$loss_reference, asymmetric_loss)
error <- tryCatch(
  invalid_native$invoke_training_loss(
    fixture$task_json$FIT_CV,
    target = 2,
    prediction = 5
  ),
  error = identity
)
stopifnot(
  inherits(error, "error"),
  grepl("failed to load DAG-ML native library", conditionMessage(error)),
  calls == 2L
)

error <- tryCatch(
  registry$invoke_training_loss(
    fixture$tasks$FIT_CV,
    target = 2,
    prediction = 5
  ),
  error = identity
)
stopifnot(
  inherits(error, "error"),
  grepl("NodeTask JSON", conditionMessage(error)),
  calls == 2L
)

failing <- dagml_local_implementation_registry()
failing$register_loss(fixture$loss_reference, function(...) stop("local failure"))
error <- tryCatch(
  failing$invoke_training_loss(
    fixture$task_json$FIT_CV,
    target = 2,
    prediction = 5
  ),
  error = identity
)
stopifnot(inherits(error, "error"), grepl("local failure", conditionMessage(error)))

error <- tryCatch(registry$to_json(), error = identity)
stopifnot(inherits(error, "error"), grepl("cannot be serialized", conditionMessage(error)))

removed <- registry$unregister_loss(fixture$loss_reference)
stopifnot(identical(removed, asymmetric_loss), registry$size() == 1L)
removed_metric <- registry$unregister_metric(fixture$metric_reference)
stopifnot(identical(removed_metric, bias_metric), registry$size() == 0L)
registry$register_metric(fixture$metric_reference, bias_metric)
registry$clear()
stopifnot(registry$size() == 0L)

File: bindings/matlab/tests/local_implementation_registry.m
function local_implementation_registry()
% Exercise MATLAB-owned loss and metric functions against native NodeTasks.

bindingRoot = fileparts(fileparts(mfilename('fullpath')));
addpath(bindingRoot);
fixturePath = fullfile(bindingRoot, 'fixtures', ...
    'matlab_local_implementations.v1.json');
fixture = jsondecode(fileread(fixturePath));

global DAGML_TEST_LOSS_CALLS;
DAGML_TEST_LOSS_CALLS = 0;
registry = dagml.LocalImplementationRegistry();
registry.registerLoss(fixture.loss_reference, @asymmetricLoss);
registry.registerMetric(fixture.metric_reference, @biasMetric);
assert(registry.count() == 2);
assert(numel(registry.descriptors()) == 2);

phases = {'FIT_CV', 'REFIT'};
for index = 1:numel(phases)
    phase = phases{index};
    [value, attestation] = registry.invokeTrainingLoss( ...
        fixture.task_json.(phase), 1, [2, 4], [5, 3]);
    assert(abs(value - 5.5) < eps);
    assert(strcmp(attestation.phase, phase));
    assert(strcmp(attestation.descriptor_fingerprint, ...
        fixture.loss_reference.implementation.descriptor_fingerprint));
    assert(isequal(attestation, ...
        fixture.tasks.(phase).required_loss_attestations));
end
assert(DAGML_TEST_LOSS_CALLS == 2);

metricValue = registry.invokeMetric( ...
    fixture.metric_reference, [2, 4], [5, 3]);
assert(abs(metricValue - 1) < eps);

assertThrows(@() registry.invokeTrainingLoss( ...
    fixture.invalid_task_json.predict, 1, 2, 5), 'FIT_CV or REFIT');
assert(DAGML_TEST_LOSS_CALLS == 2);

phaseControllers = containers.Map();
phaseControllers('controller:matlab-local') = ...
    @(controllerId, taskJson) error('dagml:test:Unexpected', ...
    'unexpected callback'); %#ok<NASGU>
assertThrows(@() dagml.executeExecutionPlanPhase( ...
    '{}', {}, 'run:matlab-phase', 1, 'FIT_CV', phaseControllers, ''), ...
    'native library');

assertThrows(@() registry.registerLoss( ...
    fixture.foreign_loss_reference, @asymmetricLoss), 'binding:matlab');

portableBuiltin = fixture.loss_reference;
portableBuiltin.implementation.portability = 'portable_builtin';
portableRegistry = dagml.LocalImplementationRegistry();
assertThrows(@() portableRegistry.registerLoss( ...
    portableBuiltin, @asymmetricLoss), 'portable_builtin');
assertThrows(@() registry.registerLoss( ...
    fixture.loss_reference, @asymmetricLoss), 'Duplicate');
assertThrows(@() registry.registerLoss( ...
    fixture.loss_reference, 42), 'function handles');
assertThrows(@() registry.resolveMetric( ...
    fixture.loss_reference), 'metric implementation descriptor');

drifted = fixture.loss_reference;
drifted.implementation.implementation_version = '2.0.0';
assertThrows(@() registry.resolveLoss(drifted), 'does not match');

assertThrows(@() registry.invokeTrainingLoss( ...
    fixture.invalid_task_json.tampered_attestation, 1, 2, 5), ...
    'requirements that do not match');
assert(DAGML_TEST_LOSS_CALLS == 2);

assertThrows(@() registry.invokeTrainingLoss( ...
    fixture.invalid_task_json.wrong_attestation_schema, 1, 2, 5), ...
    'requirements that do not match');
assert(DAGML_TEST_LOSS_CALLS == 2);

assertThrows(@() registry.invokeTrainingLoss( ...
    fixture.invalid_task_json.missing_attestation, 1, 2, 5), ...
    'requirements that do not match');
assertThrows(@() registry.invokeTrainingLoss( ...
    fixture.task_json.FIT_CV, 2, 2, 5), 'outside the active');
assert(DAGML_TEST_LOSS_CALLS == 2);

withoutNative = dagml.LocalImplementationRegistry('');
withoutNative.registerLoss(fixture.loss_reference, @asymmetricLoss);
assertThrows(@() withoutNative.invokeTrainingLoss( ...
    fixture.task_json.FIT_CV, 1, 2, 5), 'DAGML_NATIVE_LIBRARY');
assert(DAGML_TEST_LOSS_CALLS == 2);

invalidNative = dagml.LocalImplementationRegistry('/dagml/does/not/exist');
invalidNative.registerLoss(fixture.loss_reference, @asymmetricLoss);
assertThrows(@() invalidNative.invokeTrainingLoss( ...
    fixture.task_json.FIT_CV, 1, 2, 5), ...
    'Failed to load DAG-ML native library');
assert(DAGML_TEST_LOSS_CALLS == 2);

assertThrows(@() registry.invokeTrainingLoss( ...
    fixture.tasks.FIT_CV, 1, 2, 5), 'NodeTask JSON');
assert(DAGML_TEST_LOSS_CALLS == 2);

failing = dagml.LocalImplementationRegistry();
failing.registerLoss(fixture.loss_reference, @failingLoss);
assertThrows(@() failing.invokeTrainingLoss( ...
    fixture.task_json.FIT_CV, 1, 2, 5), 'local failure');

assertThrows(@() registry.toJSON(), 'cannot be serialized');
assertThrows(@() registry.saveobj(), 'cannot be serialized');

removed = registry.unregisterLoss(fixture.loss_reference);
assert(isequal(removed, @asymmetricLoss));
assert(registry.count() == 1);
removedMetric = registry.unregisterMetric(fixture.metric_reference);
assert(isequal(removedMetric, @biasMetric));
assert(registry.count() == 0);
registry.registerMetric(fixture.metric_reference, @biasMetric);
registry.clear();
assert(registry.count() == 0);
clear global DAGML_TEST_LOSS_CALLS;
end

function value = asymmetricLoss(target, prediction)
global DAGML_TEST_LOSS_CALLS;
DAGML_TEST_LOSS_CALLS = DAGML_TEST_LOSS_CALLS + 1;
difference = prediction - target;
weights = ones(size(difference));
weights(difference < 0) = 2;
value = mean(weights .* difference .^ 2);
end

function value = biasMetric(target, prediction)
value = mean(prediction - target);
end

function value = failingLoss(varargin) %#ok<INUSD,STOUT>
error('dagml:test:LocalFailure', 'local failure');
end

function assertThrows(callback, messagePart)
didThrow = false;
try
    callback();
catch exception
    didThrow = true;
    assert(~isempty(strfind(exception.message, messagePart))); %#ok<STREMP>
end
assert(didThrow);
end
